File: rust/src/leetcode/Q130_surrounded_regions.rs
/*
 * @lc app=leetcode id=130 lang=rust
 *
 * [130] Surrounded Regions
 */

// @lc code=start
impl Solution {
    fn dfs(board: &mut Vec<Vec<char>>, row: i32, col: i32) {
        static DIRECTIONS: [(i32, i32); 4] = [(-1, 0), (1, 0), (0, -1), (0, 1)];

        if board[row as usize][col as usize] != 'O' {
            return;
        }

        board[row as usize][col as usize] = 'T';

        for (dx, dy) in DIRECTIONS {
            let new_row: i32 = row + dx;
            let new_col: i32 = col + dy;

            if new_row >= 0
                && new_col >= 0
                && new_row < board.len() as i32
                && new_col < board[0].len() as i32
            {
                Self::dfs(board, new_row, new_col);
            }
        }
    }

    pub fn solve(board: &mut Vec<Vec<char>>) {
        let rows: usize = board.len();
        let cols: usize = board[0].len();

        for row in 0..rows {
            Self::dfs(board, row as i32, 0);
            Self::dfs(board, row as i32, cols as i32 - 1);
        }

        for col in 0..cols {
            Self::dfs(board, 0, col as i32);
            Self::dfs(board, rows as i32 - 1, col as i32);
        }

        for row in 0..rows {
            for col in 0..cols {
                match board[row][col] {
                    'T' => {
                        board[row][col] = 'O';
                    }
                    'O' => {
                        board[row][col] = 'X';
                    }
                    _ => {}
                }
            }
        }
    }
}
// @lc code=end

File: python/Q130_surrounded_regions.py
#
# @lc app=leetcode id=130 lang=python3
#
# [130] Surrounded Regions
#


# @lc code=start
class Solution:
    def dfs(self, row: int, col: int) -> None:
        if (
            row < 0
            or row >= self.ROWS
            or col < 0
            or col >= self.COLS
            or self.board[row][col] != "O"
        ):
            return

        self.board[row][col] = "P"

        for i, j in self.directions:
            self.dfs(row + i, col + j)

    def solve(self, board: list[list[str]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        self.ROWS: int = len(board)
        self.COLS: int = len(board[0])

        self.board: list[list[str]] = board
        self.directions: list[tuple[int, int]] = [(1, 0), (-1, 0), (0, 1), (0, -1)]

        for i in range(self.ROWS):
            self.dfs(i, 0)
            self.dfs(i, self.COLS - 1)

        for j in range(self.COLS):
            self.dfs(0, j)
            self.dfs(self.ROWS - 1, j)

        for i in range(self.ROWS):
            for j in range(self.COLS):
                match board[i][j]:
                    case "O":
                        board[i][j] = "X"
                    case "P":
                        board[i][j] = "O"


# @lc code=end
